test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java

```diff
@@ -101,75 +112,196 @@ private int runIt(String args[], PrintStream out) {
+             */
+            log.display("\nTEST #5: Trying to stop debuggee thread while suspended in Thread.sleep().");
+            waitForTestReady(5);
+            // Allow debuggee to reach Thread.sleep() first.
+            log.display("TEST #5: waiting for debuggee to sleep...");
+            while (true) {
+                int status = thrRef.status();
+                if (status == ThreadReference.THREAD_STATUS_SLEEPING ||
+                    status == ThreadReference.THREAD_STATUS_WAIT)
+                {
+                    break;
+                }
+                Thread.sleep(50);
+            }
+            log.display("TEST #5: debuggee is sleeping.");
+            try {
+                thrRef.suspend();
+                log.display("TEST #5: thread is suspended.");
+                thrRef.stop(throwableRef);
+                if (vthreadMode) {
+                    log.complain("TEST #5 FAILED: expected OpaqueFrameException was not thrown");
+                    tot_res = Consts.TEST_FAILED;
+                } else {
+                    log.display("TEST #5 PASSED: stop() call for suspended thread succeeded");
+                }
+            } catch (Throwable ue) {
+                if (vthreadMode && ue instanceof OpaqueFrameException) {
+                    log.display("TEST #5 PASSED: stop() call threw OpaqueFrameException for virtual thread");
+                } else {
+                    ue.printStackTrace();
+                    log.complain("TEST #5 FAILED: caught unexpected " + ue);
+                    tot_res = Consts.TEST_FAILED;
+                }
+            } finally {
+                log.display("TEST #5: resuming thread.");
+                thrRef.resume();
+            }
+            log.display("TEST #5: all done.");
         } catch (Exception e) {
             e.printStackTrace();
             log.complain("TEST FAILURE: caught unexpected exception: " + e);
             tot_res = Consts.TEST_FAILED;
         } finally {
-// Finish the test
-            // force an method to exit
-            if (objRef != null && doExit != null) {
+            // Force the debuggee out of both loops
+            if (objRef != null && stopLoop1 != null && stopLoop2 != null) {
                 try {
-                    objRef.setValue(doExit, vm.mirrorOf(true));
-                } catch(Exception sve) {
+                    objRef.setValue(stopLoop1, vm.mirrorOf(true));
+                    objRef.setValue(stopLoop2, vm.mirrorOf(true));
+                } catch (Exception sve) {
                     sve.printStackTrace();
+                    tot_res = Consts.TEST_FAILED;
                 }
             }
         }
 
         return quitDebuggee();
     }
 
+    private void waitForTestReady(int testNum) {
+        log.display("TEST #" + testNum + ": waiting for test ready...");
+        IntegerValue ival;
+        do {
+            ival = (IntegerValue)mainClass.getValue(mainClass.fieldByName("testNumReady"));
```

Comment:
Do we need a sleep at each iteration?

Reply:
I intentionally did not do one since `getValue()` is fairly slow and involves sending and receiving a JDWP packet. It didn't seem worth the extra noise of a sleep call and the needed exception handling.

Reply:
Okay, thanks.